core/domain/translation_domain.py

```diff
@@ -0,0 +1,118 @@
+
+from __future__ import absolute_import  # pylint: disable=import-only-modules
+from __future__ import unicode_literals  # pylint: disable=import-only-modules
+
+import python_utils
+import utils
+
+
+class MachineTranslatedText(python_utils.OBJECT):
+    """Domain object for machine translation of exploration content."""
+
+    def __init__(
+            self, source_language_code, target_language_code, source_text,
+            translated_text):
+        """Initializes a MachineTranslatedText domain object.
+
+        Args:
+            source_language_code: str. The language code for the source text
+                language. Must be different from target_language_code.
+            target_language_code: str. The language code for the target
+                translation language. Must be different from
+                source_language_code.
+            source_text: str. The untranslated source text.
+            translated_text: str. The machine generated translation of the
+                source text into the target language.
+        """
+        self.source_language_code = source_language_code
+        self.target_language_code = target_language_code
+        self.source_text = source_text
+        self.translated_text = translated_text
+
+    def validate(self):
+        """Validates properties of the MachineTranslatedText.
+
+        Raises:
+            ValidationError. One or more attributes of the MachineTranslatedText
+                are invalid.
+        """
+        if not isinstance(self.source_language_code, python_utils.BASESTRING):
+            raise utils.ValidationError(
+                'Expected source_language_code to be a string, received %s' %
+                self.source_language_code
+            )
+        # TODO(#12341): Tidy up this logic once we have a canonical list of
+        # language codes.
+        if not utils.is_supported_audio_language_code(
+                self.source_language_code
+            ) and not utils.is_valid_language_code(
+                self.source_language_code
+            ):
+            raise utils.ValidationError(
+                'Invalid source language code: %s' % self.source_language_code)
+
+        if not isinstance(self.target_language_code, python_utils.BASESTRING):
+            raise utils.ValidationError(
+                'Expected target_language_code to be a string, received %s' %
+                self.target_language_code
+            )
+        # TODO(#12341): Tidy up this logic once we have a canonical list of
+        # language codes.
+        if not utils.is_supported_audio_language_code(
+                self.target_language_code
+            ) and not utils.is_valid_language_code(
+                self.target_language_code
+            ):
+            raise utils.ValidationError(
+                'Invalid target language code: %s' % self.target_language_code)
+
+        if self.source_language_code is self.target_language_code:
```

Comment:
Probably should use `==` here, since you're comparing equality of two strings which isn't an "is-a" thing.

Reply:
Fixed. I'm surprised `test_validate_with_same_source_target_language_codes_raises` didn't fail.